File: web/src/app/core/utils/client-location.ts
/** Optional GPS snapshot for audit metadata on sensitive API posts. */
export interface ClientLocationSnapshot {
  client_latitude: string;
  client_longitude: string;
  client_location_accuracy_m?: number;
}

/**
 * Returns current browser location only when geolocation permission is already
 * granted. Never prompts and never throws — resolves null otherwise.
 */
export async function optionalClientLocation(
  timeoutMs = 4000,
): Promise<ClientLocationSnapshot | null> {
  if (typeof navigator === 'undefined' || !navigator.geolocation) {
    return null;
  }

  // Skip unless permission is already granted (no prompt on payment/booking actions).
  const permissions = navigator.permissions;
  if (!permissions?.query) {
    return null;
  }
  try {
    const status = await permissions.query({
      name: 'geolocation' as PermissionName,
    });
    if (status.state !== 'granted') {
      return null;
    }
  } catch {
    return null;
  }

  try {
    const position = await new Promise<GeolocationPosition>((resolve, reject) => {
      navigator.geolocation.getCurrentPosition(resolve, reject, {
        enableHighAccuracy: false,
        timeout: timeoutMs,
        maximumAge: 60_000,
      });
    });
    const { latitude, longitude, accuracy } = position.coords;
    if (!Number.isFinite(latitude) || !Number.isFinite(longitude)) {
      return null;
    }
    const snapshot: ClientLocationSnapshot = {
      client_latitude: latitude.toFixed(6),
      client_longitude: longitude.toFixed(6),
    };
    if (Number.isFinite(accuracy) && accuracy >= 0) {
      snapshot.client_location_accuracy_m = accuracy;
    }
    return snapshot;
  } catch {
    return null;
  }
}

/** Merge optional GPS fields into an API body object. */
export async function withOptionalClientLocation(
  body: Record<string, unknown> = {},
): Promise<Record<string, unknown>> {
  const loc = await optionalClientLocation();
  if (!loc) {
    return body;
  }
  return { ...body, ...loc };
}

File: mobile/lib/core/utils/client_location.dart
import 'package:location/location.dart';

/// Optional GPS coordinates for audit metadata on sensitive API posts.
class ClientLocationSnapshot {
  const ClientLocationSnapshot({
    required this.latitude,
    required this.longitude,
    this.accuracyMeters,
  });

  final double latitude;
  final double longitude;
  final double? accuracyMeters;

  Map<String, dynamic> toJson() => {
        'client_latitude': latitude.toStringAsFixed(6),
        'client_longitude': longitude.toStringAsFixed(6),
        if (accuracyMeters != null)
          'client_location_accuracy_m': accuracyMeters,
      };
}

/// Returns current location when permission is already granted.
/// Never prompts and never throws — returns null when unavailable or denied.
Future<ClientLocationSnapshot?> optionalClientLocation() async {
  try {
    final location = Location();
    final serviceEnabled = await location.serviceEnabled();
    if (!serviceEnabled) return null;

    final permission = await location.hasPermission();
    if (permission != PermissionStatus.granted &&
        permission != PermissionStatus.grantedLimited) {
      return null;
    }

    final data = await location.getLocation();
    final lat = data.latitude;
    final lng = data.longitude;
    if (lat == null || lng == null) return null;
    return ClientLocationSnapshot(
      latitude: lat,
      longitude: lng,
      accuracyMeters: data.accuracy,
    );
  } on Object {
    return null;
  }
}

/// Merge optional GPS fields into an API body map.
Future<Map<String, dynamic>> withOptionalClientLocation(
  Map<String, dynamic> body,
) async {
  final loc = await optionalClientLocation();
  if (loc == null) return body;
  return {...body, ...loc.toJson()};
}
